absl/absl_test.cpp

```diff
@@ -24,36 +24,153 @@
 namespace nda {
 
 TEST(absl_stringify_interval) {
-  interval<> i0;
   {
+    interval<> i0;
     const std::string s = absl::StrFormat("%v", i0);
     ASSERT(s == "[0, 1)");
   }
 
-  interval<> i1(-6, 11);
   {
+    interval<> i1(-6, 11);
     const std::string s = absl::StrFormat("%v", i1);
     ASSERT(s == "[-6, 5)");
   }
 
-  interval<3> i2;
   {
+    interval<3> i2;
     const std::string s = absl::StrFormat("%v", i2);
     ASSERT(s == "[3, 4)");
   }
 
-  interval<4> i3;
-  i3.set_extent(6);
   {
+    interval<4> i3;
+    i3.set_extent(6);
     const std::string s = absl::StrFormat("%v", i3);
     ASSERT(s == "[4, 10)");
   }
 
-  interval<-5, -12> i4;
   {
+    interval<-5, -12> i4;
     const std::string s = absl::StrFormat("%v", i4);
     ASSERT(s == "[-5, -17)");
   }
 }
 
+TEST(absl_stringify_dim) {
+  // Fully dynamic but unset.
+  {
+    dim<> d;
+    const std::string s = absl::StrFormat("%v", d);
+    ASSERT(s == "dim(0, 0, -9)");
+  }
+
+  {
+    dim<> d(/*extent=*/640);
+    const std::string s = absl::StrFormat("%v", d);
+    ASSERT(s == "dim(0, 640, -9)");
+  }
+
+  {
+    dim<> d(/*min=*/35, /*extent=*/640);
+    const std::string s = absl::StrFormat("%v", d);
+    ASSERT(s == "dim(35, 640, -9)");
+  }
+
+  {
+    dim<> d(/*min=*/77, /*extent=*/480, /*stride=*/2);
+    const std::string s = absl::StrFormat("%v", d);
+    ASSERT(s == "dim(77, 480, 2)");
+  }
+
+  // Statically set min/extent/stride.
+  {
+    dim</*min=*/3> d;
+    const std::string s = absl::StrFormat("%v", d);
+    ASSERT(s == "dim(3, 0, -9)");
+  }
+
+  {
+    dim</*min=*/-4, /*extent=*/5> d;
+    const std::string s = absl::StrFormat("%v", d);
+    ASSERT(s == "dim(-4, 5, -9)");
+  }
+
+  {
+    dim</*min=*/10, /*extent=*/11, /*stride=*/-1> d;
+    const std::string s = absl::StrFormat("%v", d);
+    ASSERT(s == "dim(10, 11, -1)");
+  }
+}
+
+TEST(absl_stringify_shape) {
+  // Fully dynamic but unset.
+  {
+    shape_of_rank<0> sh;
+    const std::string s = absl::StrFormat("%v", sh);
+    ASSERT(s == "shape<0>()");
+  }
+
+  // Fully dynamic but unset.
+  {
+    shape_of_rank<1> sh;
+    const std::string s = absl::StrFormat("%v", sh);
+    ASSERT(s == "shape<1>(dim(0, 0, -9))");
+  }
+
+  {
+    shape_of_rank<3> sh;
+    const std::string s = absl::StrFormat("%v", sh);
+    ASSERT(s == "shape<3>(dim(0, 0, -9), dim(0, 0, -9), dim(0, 0, -9))");
```

Comment:
Seeing this, I wonder if we should try to think of a more concise syntax. It already loses information and isn't copy-pastable into C++ code (kinda useful for debugging sometimes BTW), so maybe straying further still is a good idea. For example, maybe shapes should just render like lists.

Reply:
Hm, thinking about this some more - it really is another struct of arrays vs array of structs problem. Do you want to look at all the mins at once? Or (min, extent, stride) of each dim before going to the next?

Re: compile time types - IIRC, you mentioned that we should probably drop those. Most of the time, you don't care about those and only care about logging the runtime extents. I agree - for the simpler `AbslStringify` version, we want a simpler thing.

I initially directly rendered a dim as a list but it was confusing. You end up with something like `[(0, 640, 1), (0, 480, 640), (0, 3, 307200))]` without knowing that each element is a `dim`.

Once I added `dim`, then it made sense to add `shape`. But it became irritating to count how many `dim`s there are, so that's how I ended up with the version we have now.

Let me know if there's something else you have in mind - I kinda like this version right now.

Re: customizing the formatting to include compile-time information. How about as a TODO for the [AbslFormatConvert](https://abseil.io/docs/cpp/guides/format#abslformatconvert) version. We can pick a flag (`%d` isn't bad), precision specifier or some conversion character to control whether compile time information is emitted.